### ISSUE TITLE
v7 regression: installation time is super slow

### ISSUE THREAD

**Issue by AviVahl** [tannerlinsley/react-table#2215]
**Describe the bug (required)**
We've recently upgraded from v6 to v7. Since that upgrade, our yarn installation awaits `react-table`'s `@scarf/scarf` dependency for a very long time (over 50s):
```
$ yarn
yarn install v1.22.4
[1/4] Resolving packages...
[2/4] Fetching packages...
info fsevents@1.2.12: The platform "linux" is incompatible with this module.
info "fsevents@1.2.12" is an optional dependency and failed compatibility check. Excluding it from installation.
info fsevents@2.1.3: The platform "linux" is incompatible with this module.
info "fsevents@2.1.3" is an optional dependency and failed compatibility check. Excluding it from installation.
[3/4] Linking dependencies...
[4/4] Building fresh packages...
[-/7] ⠠ waiting...
[-/7] ⠠ waiting...
[-/7] ⠐ waiting...
[4/7] ⠐ @scarf/scarf
[-/7] ⠐ waiting...
```

It finally ends after over a minute:
`Done in 64.09s.`

If I simply force v6 using yarn's "resolutions":
```
  "resolutions": {
    "react-table": "6.11.5"
  },
```
Installation is finished in `9.82s.`
Forcing v6 obviously breaks at runtime, but clearly shows the install time speed difference.

Should be noted the times above are with an up-to-date `yarn.lock` file and no `node_modules` (`git clean -fdx` between installation runs).

**Provide an example via Codesandbox! (required)**
Irrelevant when discussion installation time issues, I guess?

**Steps To Reproduce (required)**
It appears this behaviour is best seen in repositories with *many* dependencies. When I tested v7 on https://github.com/wixplosives/sample-monorepo, I could see a slowdown, but it wasn't as severe.

**Expected behavior (Recommended)**
Stay fast? :)

**Desktop (please complete the following information):**

- OS: Fedora 31
- Node: v12.16.2
- npm: 6.14.4
- yarn: 1.22.4

**Comment by AviVahl** [comment]
`@scarf/scarf/report.js` executes `npm ls @scarf/scarf --json --long`, which takes a *long time* in our repository. I mean, we're not even using npm (repository is yarn workspaces based).

Without going into my personal thoughts about the invasiveness of this `postinstall` script, this is not an ideal developer experience.

**Comment by aviaviavi** [comment]
Hi @AviVahl 👋, author of Scarf here. Apologies here for the issue you encountered, and thanks for reporting. The fix has been pushed out. 

It's understandable that you find the postinstall hook to be invasive. You can opt out of the analytics via package.json configuration or environment variable, as shown here: https://github.com/scarf-sh/scarf-js#as-a-user-of-a-package-using-scarf-how-can-i-opt-out-of-analytics. Please feel free to reach out on scarf-js's github repo if you have any questions or want to discuss this further!

**Comment by AviVahl** [comment]
heya @aviaviavi 

Thanks for addressing this with the timeout. I guess I'll need to wait for react-table to merge/release #2183 to get that fix.

I've got several questions regarding opting out:
1. In a yarn `workspace` (repository containing several packages), does it matter if I put it in the root `package.json` once? or do I need to put it in every package depending on `react-table`? Or both, in case you sometimes run `yarn` from the root directory, and at other times from a package's directory.
2. For a components library depending on `react-table` (which depends on scarf), would my opting out also opt out for users consuming my library via npm? or would each need to opt out on his own package?

**Comment by aviaviavi** [comment]
> does it matter if I put it in the root package.json once? or do I need to put it in every package depending on react-table? 

Because Scarf is currently relying on `npm ls` to identify its direct dependent package(s) in the dependency tree, it won't see the top-level workspace `package.json` when `yarn` is run from one of the individual sub-packages. This is definitely fixable though, so we will push another update to Scarf so it is workspace-aware to make that better. Thanks for bringing that up. 

In the meantime, it might be easier to take the environment variable opt-out approach at this time, but updating the packages individually would certainly work too. 

> For a components library depending on react-table (which depends on scarf), would my opting out also opt out for users consuming my library via npm? or would each need to opt out on his own package?

Right now, the user would have to do the opting out. Scarf checks the setting of the package directly depending on Scarf, and the user's settings. The behavior you're describing sounds reasonable to me, but it probably warrants some discussion on Scarf's side. I'll open that discussion for how this ought to be handled on Scarf's repo.

**Comment by AviVahl** [comment]
Thanks for your response.

Unfortunately, the environment variables approach won't work for our case, as this is a repository used by over 25 developers and is connected to 3 CI systems. Chasing down all the places and people to reconfigure their machines sounds like something I would try to avoid.

I'll check the `package.json` opting out mode.

Just noting, if I won't be able to disable this for users of a package that uses `react-table`, I may have to fork `react-table` itself. This is not something I can force push to my users.

**Comment by tannerlinsley** [comment]
Hey Avi (Vahl ;), I am pretty sure we can figure something out here, so don't go bailing just yet. Scarf is definitely still fresh and Avi (Press ;) has been working with me to polish it up as we go, but it's definitely not worth it to me to have you forking the repo to get around it.

Avi (Press), what else can we do right now the interim to help Avi (Vahl) get around this?

**Comment by tannerlinsley** [comment]
If it means publishing new versions without Scarf until we can figure it out I'm willing to do that.

**Comment by aviaviavi** [comment]
Understood and absolutely agree that we wouldn't want anyone needing to fork `react-table` (or any package) to get around this limitation. I'll update Scarf ASAP so it respects all intermediate opt-out settings in this way.

**Comment by AviVahl** [comment]
Thank you both for your prompt replies.

Trust me, forking is a "last resort" kind of option which I am always trying to avoid. Syncing changes to upstream is a headache and it gets even trickier with TypeScript and how `@types/<package-name>` are resolved. I would rather find a solution that works for the official `react-table` package.

Respecting intermediate opt-out sounds like something that could resolve the issue for us. I'll update as soon as you have any fix released.

On a side note: I really appreciate the work done on `react-table`, and find it super useful. Kudos for it.

**Comment by tannerlinsley** [comment]
Thanks Avi Vahl

I had no idea that Wix used it! I'd love that logo up there, too. You guys ever think of sponsoring?

**Comment by AviVahl** [comment]
I can't speak for the rest of Wix (many developers; tons of repositories), but my team started using `react-table` inside a product we're working on. It's not public yet, nor do I know whether this specific feature (the one that uses `react-table`) will end up in the GA version, so can't promise anything.

I'll ask around about possible sponsorship. TBH, I'm not sure of anything that's not code-related... :)

**Comment by aviaviavi** [comment]
Update: Scarf 1.0.4 is released and enables any intermediate package in a dependency chain leading to Scarf to disable analytics. By adding that setting to the `package.json` of any packages you distribute on npm, Scarf will be disabled for all users and packages depending on yours. #2183 has been updated to pull this newest version.

The discussed change for yarn workspaces is on the way too!

**Comment by AviVahl** [comment]
Awesome! I'll wait for #2183 to merge/release and test.

**Comment by AviVahl** [comment]
@aviaviavi btw, I saw you've also released a new pre-1.0.0 release. It appears to have the timeout, and I no longer see the slowdown that was reported.

Regarding un-subscribing my own library users, is that included in that pre-1.0.0 version as well, or do I need to wait for above PR to merge to get that behaviour?

I'm asking because it's difficult for me to determine which branch of the flow scarf picked up, as yarn completely hides the output of `report.js`.

**Comment by aviaviavi** [comment]
Good to hear that change got picked up and took care of the slowdown. You are correct that it is indeed going to be hitting that timeout for you, as the operation that caused the slowness is needed just to even calculate opt-in/out. 

For un-subscribing your own library users: That patch has not yet been merged to the 0.1.x release yet, but it will be shortly. I'll be sure to send an update here when it goes out.

Because of yarn's current behavior of suppressing the output of transitive dependencies, additional changes are on the way to automatically disable the analytics when the end-user is installing via yarn. We want to only report these analytics when we can transparently inform the user it is happening.

**Comment by AviVahl** [comment]
@aviaviavi

Thanks for the clarifications and addressing the various concerns.

I'm closing this issue (slowdown is resolved) and will follow #2183 and `react-table` releases to get the un-subscription behavior. And in the future, once it behaves properly on `yarn` (either by auto disabling or any other alternatives you come up with), I wouldn't even mind leaving it with the default behavior so that @tannerlinsley gets the statistics he wants.

Have a great week!

**Comment by aviaviavi** [comment]
Sure thing! That all sounds great, and your support there is very appreciated. Will keep you posted with these updates.

You too! 😄
